src/udev/memory_id/dmidecode.c

```diff
@@ -0,0 +1,768 @@
+
+        if (s == 0)
+                return "Not Specified";
+
+        bp = _dmi_string(dm, s);
+        if (bp == NULL)
+                return bad_index;
+
+        return bp;
+}
+
+/* shift is 0 if the value is in bytes, 1 if it is in kilobytes */
+static void dmi_print_memory_size(const char *attr_prefix, const char *attr_suffix, int slot_num, u64 code, int shift)
+{
+        uint64_t capacity;
+
+        capacity = code.h;
+        capacity = capacity << 32 | code.l;
+        if (shift == 1)
+                capacity = capacity << 10;
+
+        if (slot_num >= 0)
+                printf("%s_%u_%s=%"PRIu64"\n", attr_prefix, slot_num, attr_suffix, capacity);
+        else
+                printf("%s_%s=%"PRIu64"\n", attr_prefix, attr_suffix, capacity);
+}
+
+/*
+ * 7.17 Physical Memory Array (Type 16)
+ */
+
+static const char *dmi_memory_array_location(uint8_t code)
+{
+        /* 7.17.1 */
+        static const char *location[] = {
+                "Other", /* 0x01 */
+                "Unknown",
+                "System Board Or Motherboard",
+                "ISA Add-on Card",
+                "EISA Add-on Card",
+                "PCI Add-on Card",
+                "MCA Add-on Card",
+                "PCMCIA Add-on Card",
+                "Proprietary Add-on Card",
+                "NuBus" /* 0x0A */
+        };
+        static const char *location_0xA0[] = {
+                "PC-98/C20 Add-on Card", /* 0xA0 */
+                "PC-98/C24 Add-on Card",
+                "PC-98/E Add-on Card",
+                "PC-98/Local Bus Add-on Card",
+                "CXL Flexbus 1.0" /* 0xA4 */
+        };
+
+        if (code >= 0x01 && code <= 0x0A)
+                return location[code - 0x01];
+        if (code >= 0xA0 && code <= 0xA4)
+                return location_0xA0[code - 0xA0];
+        return out_of_spec;
+}
+
+static const char *dmi_memory_array_ec_type(uint8_t code)
+{
+        /* 7.17.3 */
+        static const char *type[] = {
+                "Other", /* 0x01 */
+                "Unknown",
+                "None",
+                "Parity",
+                "Single-bit ECC",
+                "Multi-bit ECC",
+                "CRC" /* 0x07 */
+        };
+
+        if (code >= 0x01 && code <= 0x07)
+                return type[code - 0x01];
+        return out_of_spec;
+}
+
+/*
+ * 7.18 Memory Device (Type 17)
+ */
+
+static void dmi_memory_device_width(const char *attr_prefix, const char *attr_suffix, unsigned int slot_num, uint16_t code)
+{
+        /*
+         * If no memory module is present, width may be 0
+         */
+        if (code != 0xFFFF && code != 0)
+                printf("%s_%u_%s=%u\n", attr_prefix, slot_num, attr_suffix, code);
+}
+
+static void dmi_memory_device_size(unsigned int slot_num, uint16_t code)
+{
+        if (code == 0) {
+                printf("MEMORY_DEVICE_%u_PRESENT=0\n", slot_num);
+                return;
+        } else if (code == 0xFFFF) {
+                return;
+        } else {
```

Comment:
doesn't matter, but we usually don't place {} around single-line if blocks

Reply:
I find it looks exceedingly weird when one conditional has them, but not the others. Let me know if you still want that after checking the last coding style changes.